taf/testlib/virtual_env.py

```diff
@@ -304,26 +371,24 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
                 'start': '{}.100'.format(allocation_prefix),
                 'end': '{}.254'.format(allocation_prefix)}],
             'gateway_ip': net_ip.ip,
-            'enable_dhcp': False
+            'enable_dhcp': False,
         }
         self._create_subnet(public_network['id'], **subnet_kwargs)
 
         _net_cfg.public_network_id = public_network['id']
 
-        router_kwargs = {
-            'routers_client': routers_client,
-            'network_id': _net_cfg.public_network_id,
-            'tenant_id': tenant_id,
-            'enable_snat': True
-        }
-        public_router = self.create_router(**router_kwargs)
-        assert public_router
-        _net_cfg.public_router_id = public_router['id']
-
-        if _net_cfg.public_network_id and _net_cfg.public_router_id:
-            return True
+        if create_external_router:
+            router_kwargs = {
+                'routers_client': routers_client,
+                'network_id': _net_cfg.public_network_id,
+                'tenant_id': tenant_id,
+                'enable_snat': True,
+            }
+            public_router = self.create_router(**router_kwargs)
+            assert public_router
+            _net_cfg.public_router_id = public_router['id']
 
-        return False
+        return  _net_cfg.public_network_id and (_net_cfg.public_router_id or not create_external_router)
```

Comment:
Looking at this again, would it make more sense to put the `not create_external_router` clause before `_net_cfg.public_router_id`?  Because the latter could raise an AttributeError even when the result should be True.

```return _net_cfg.public_network_id and (not create_external_router or _net_cfg.public_router_id)```